lib/vsc/modules/cache.py

```diff
@@ -0,0 +1,293 @@
+
+            # track the default
+            #   the keys also can be used as list of all clusters that have the software
+            softdefault = soft.setdefault(DEFAULTKEY, {})
+            for cluster in clusters:
+                tmpdefault = softdefault.setdefault(cluster, default)
+                if tmpdefault != default:
+                    # typically due to modulepath ordering
+                    LOGGER.debug("Already found default for %s for cluster %s: found %s, new %s",
+                                 name, cluster, tmpdefault, default)
+    return softmap
+
+
+def get_json_filename():
+    """Return the filename of the JSON data"""
+    config = get_lmod_conf()
+    return os.path.join(config['dir'], JSON_MODULEMAP_FILENAME)
+
+
+def write_json(clustermap, softmap, filename=None):
+    """Write JSON with cluster and software map"""
+    if filename is None:
+        filename = get_json_filename()
+
+    # TODO: make atomic
+    with open(filename, 'w') as outfile:
+        json.dump({
+            MAIN_CLUSTERS_KEY: clustermap,
+            MAIN_SOFTWARE_KEY: softmap,
+        }, outfile)
+        LOGGER.debug("Wrote %s", filename)
+
+
+def read_json(filename=None):
+    """Read JSON and return cluster and software map"""
+    if filename is None:
+        filename = get_json_filename()
+    with open(filename) as outfile:
+        data = json.load(outfile)
+        LOGGER.debug("Read %s", filename)
+
+    return data[MAIN_CLUSTERS_KEY], data[MAIN_SOFTWARE_KEY]
+
+
+def software_cluster_view(softmap=None):
+    """
+    Return a dict with list of software versions per cluster.
+    First version is the default, the remainder is sorted
+    """
+    if softmap is None:
+        _, softmap = read_json()
+
+    clview = {}
+    for name, vdata in softmap.items():
+        defaults = vdata.pop(DEFAULTKEY)
+
+        # the keys have all clusters that have software name
+        for cluster in defaults.keys():
+            # create nested structure cluster -> name -> versions
+            clsoft = clview.setdefault(cluster, {})
+            clsoft.setdefault(name, [])
+
+        for version, clusters in vdata.items():
+            for cluster in clusters:
+                clview[cluster][name].append(version)
+
+        for cluster, default in defaults.items():
+            versions = sort_recent_versions(clview[cluster][name])
+            clview[cluster][name] = versions
+            try:
+                versions.remove(default)
+            except ValueError as e:
+                LOGGER.raiseException("Unable to remove default %s from versions for %s cluster %s: %s" %
+                                      (default, versions, name, cluster, e))
+
+            versions.insert(0, default)  # default first
+
+    return clview
+
+
+def convert_lmod_cache_to_json():
+    """Main conversion of Lmod lua cache to cluster and software mapping in JSON"""
+    # you really don't want this in debug
```

Comment:
You mean there shouldn't be a debug log statement including the cache contents, I think?

If so, please clarify the comment :)

Reply:
i removed it, if you run it in debug, you'll get the output (but you probably don't want it)